### PR TITLE
Skip ASN.1 SetOf sorting if nothing to sort

### PR DESCRIPTION
**Pull request by vcsjones**
Many ASN.1 SET OFs that get encoded end up containing only one thing, such as RDNs, SignerInfos in CMS, etc.

The current ASN.1 DER implementation sorts SET OF (good) but it does this sorting work even if there is only one thing in the SET OF. This results in unnecessary copying of data. This change skips the sorting work if there nothing to sort.

This is easily visible in a (pretty unrealistic) benchmark of writing a SET OF that contains a single OCTET STRING that is 1 GB.

|        Method |        Job | Toolchain |     Mean |   Error |  StdDev | Ratio | Allocated |
|-------------- |----------- |---------- |---------:|--------:|--------:|------:|----------:|
| WriteSetOfDER | Job-RWHAMN |        pr | 239.3 ms | 1.85 ms | 1.73 ms |  0.63 |      2 GB |
| WriteSetOfDER | Job-YJKRNM |      main @ f2d65eb | 379.0 ms | 3.09 ms | 2.74 ms |  1.00 |      3 GB |

A more typical usage will still see improvements from skipping a rental and two `BlockCopy` invocations.

### CODE REVIEW

File: src/libraries/System.Formats.Asn1/src/System/Formats/Asn1/AsnWriter.cs [hunk 1]
```diff
@@ -621,6 +621,12 @@ private static void SortContents(byte[] buffer, int start, int end)
 
             Debug.Assert(pos == end);
 
+            // If there is only one thing in the contents, we have no sorting work to do and it's already sorted.
+            if (positions.Count == 1)
```
Comment (stephentoub) [inline src/libraries/System.Formats.Asn1/src/System/Formats/Asn1/AsnWriter.cs:625]: Count can't be 0?

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Formats.Asn1/src/System/Formats/Asn1/AsnWriter.cs (hunk 1 above)
Comment (bartonjs) [reply src/libraries/System.Formats.Asn1/src/System/Formats/Asn1/AsnWriter.cs:625]: Count can only be 0 if the data set is empty, which was an earlier exit.

Comment (bartonjs) [inline src/libraries/System.Formats.Asn1/src/System/Formats/Asn1/AsnWriter.cs:625]: If it's enough of a case that it's worth an early exit, then we should probably rewrite 611-622 to something like

```C#
ReadOnlyMemory<byte> encoded = reader.ReadEncodedValue();

if (!reader.HasData)
{
    return;
}

List<(int, int)> positions = new List<(int, int)>
{
    (pos, encoded.Length),
};

pos += encoded.Length;

// the current loop
```

that way we don't create the List object if we're going to early exit.

Comment (vcsjones) [reply src/libraries/System.Formats.Asn1/src/System/Formats/Asn1/AsnWriter.cs]: @bartonjs seems reasonable.

**Review by bartonjs — APPROVED** [review]

**Pull request merged**